cli/flox-test-utils/src/proptest.rs

```diff
@@ -39,11 +69,126 @@ pub fn empty_btree_map_alphanum_keys(
         0..max_keys,
     )
 }
+
+/// Produces `Vec<String>` instances with limited sizes for performance reasons.
+pub fn vec_of_strings(
+    string_max_size: usize,
+    max_elements: usize,
+) -> impl Strategy<Value = Vec<String>> {
+    prop_vec(alphanum_string(string_max_size), 0..=max_elements)
+}
+
+/// Produces `Option<Vec<String>>` instances with limited sizes for performance
+/// reasons
+pub fn optional_vec_of_strings(
+    string_max_size: usize,
+    max_elements: usize,
+) -> impl Strategy<Value = Option<Vec<String>>> {
+    option::of(prop_vec(alphanum_string(string_max_size), 1..=max_elements))
+}
+
+/// A container for randomly generated maps with overlapping keys
+#[derive(Debug, Clone)]
+pub struct OverlappingMaps<T> {
+    pub unique_keys_map1: Vec<String>,
+    pub unique_keys_map2: Vec<String>,
+    pub duplicate_keys: Vec<String>,
+    pub map1: BTreeMap<String, T>,
+    pub map2: BTreeMap<String, T>,
+}
+
+/// Produces two maps whose keys overlap.
+///
+/// The return values are the two maps whose keys overlap and the overlapping keys.
+/// Note that this returns a `BoxedStrategy` purely for the sake of not killing
+/// `rust-analyzer` with the horendous type that would otherwise be generated as
+/// the return type for this function.
+pub fn btree_maps_overlapping_keys<T: Arbitrary>(
+    key_max_size: usize,
+    max_keys: usize,
+) -> BoxedStrategy<OverlappingMaps<T>> {
+    (0..max_keys, 0..max_keys, 0..max_keys)
+        .prop_flat_map(move |(n_keys_map1, n_keys_map2, n_dupes)| {
+            (
+                // Keys unique to map 1
+                prop_vec(alphanum_string(key_max_size), n_keys_map1..=n_keys_map1),
+                // Keys unique to map 2
+                prop_vec(alphanum_string(key_max_size), n_keys_map2..=n_keys_map2),
+                // Keys duplicated between the maps
+                prop_vec(alphanum_string(key_max_size), n_dupes..=n_dupes),
+                // Values for the unique keys in map 1
+                prop_vec(any::<T>(), n_keys_map1..=n_keys_map1),
+                // Values for the unique keys in map 2
+                prop_vec(any::<T>(), n_keys_map2..=n_keys_map2),
+                // Values for the duplicate keys in map 1
+                prop_vec(any::<T>(), n_dupes..=n_dupes),
+                // Values for the duplicate keys in map 2
+                prop_vec(any::<T>(), n_dupes..=n_dupes),
+            )
+        })
+        .prop_filter(
+            "ensure no pre-existing overlap",
+            |(
+                keys_map1,
+                keys_map2,
+                keys_dup,
+                _vals_map1,
+                _vals_map2,
+                _vals_dup_map1,
+                _vals_dup_map2,
+            )| {
+                // Ensure that there's no pre-existing overlap with the duplicate keys
+                for key in keys_map1.iter() {
+                    if keys_dup.contains(key) {
+                        return false;
```

Comment:
question nonblocking: it's okay if all keys in map1 are in keys_dup and get filtered right? That would just be the case where there are no unique keys in map1?

Reply:
I'm not sure if I'm understanding the generator correctly, but it seems like if it adds `whoops` to both `keys_map1` and `keys_dup`, it will filter it out of `keys_map1`. That could leave `keys_map1` completely empty. I think that's fine because that's a good case to test, and other cases will generate a non-empty `keys_map1`, but I was just checking my understanding

Reply:
> it will filter it out of `keys_map1`

I'm not sure if you misspoke here, but that's not what this method does. This method rejects randomly generated samples based on the predicate function, it doesn't modify the inputs.

Reply:
Ah that's what I was misunderstanding. Thanks 👍